geode-core/src/main/java/org/apache/geode/internal/cache/control/RestoreRedundancyBuilderImpl.java

```diff
@@ -0,0 +1,177 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more contributor license
+ * agreements. See the NOTICE file distributed with this work for additional information regarding
+ * copyright ownership. The ASF licenses this file to You under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance with the License. You may obtain a
+ * copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software distributed under the License
+ * is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
+ * or implied. See the License for the specific language governing permissions and limitations under
+ * the License.
+ */
+package org.apache.geode.internal.cache.control;
+
+import java.util.List;
+import java.util.Set;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.concurrent.atomic.AtomicBoolean;
+import java.util.stream.Collectors;
+
+import org.apache.geode.cache.RegionDestroyedException;
+import org.apache.geode.cache.control.RestoreRedundancyBuilder;
+import org.apache.geode.cache.control.RestoreRedundancyResults;
+import org.apache.geode.cache.partition.PartitionRebalanceInfo;
+import org.apache.geode.internal.cache.InternalCache;
+import org.apache.geode.internal.cache.PartitionedRegion;
+import org.apache.geode.internal.cache.partitioned.PartitionedRegionRebalanceOp;
+import org.apache.geode.internal.cache.partitioned.rebalance.RestoreRedundancyDirector;
+import org.apache.geode.logging.internal.log4j.api.LogService;
+
+class RestoreRedundancyBuilderImpl implements RestoreRedundancyBuilder {
+
+  private final InternalCache cache;
+  private final InternalResourceManager manager;
+  private Set<String> includedRegions;
+  private Set<String> excludedRegions;
+  private boolean shouldReassign = true;
+  private ScheduledExecutorService executor;
+
+  public RestoreRedundancyBuilderImpl(InternalCache cache) {
+    this.cache = cache;
+    this.manager = cache.getInternalResourceManager();
+    this.executor = this.manager.getExecutor();
+  }
+
+  @Override
+  public RestoreRedundancyBuilder includeRegions(Set<String> regions) {
+    this.includedRegions = regions;
+    return this;
+  }
+
+  @Override
+  public RestoreRedundancyBuilder excludeRegions(Set<String> regions) {
+    this.excludedRegions = regions;
+    return this;
+  }
+
+  @Override
+  public RestoreRedundancyBuilder setReassignPrimaries(boolean shouldReassign) {
+    this.shouldReassign = shouldReassign;
+    return this;
+  }
+
+  @Override
+  public CompletableFuture<RestoreRedundancyResults> start() {
```

Comment:
If child colocated regions are automatically rejected, then if we explicitly include a child region in the restore redundancy operation but not its parent, no work will be done. This could make life difficult for users who might not know which region in a colocation group is the parent, so the behaviour implemented here is to restore redundancy to all regions in the colocation group if any one of them is included, regardless of whether it's the parent or not.

The missing logic is also not strictly needed here, since if we call `execute()` on a `PartitionedRegionRebalanceOp` for a colocated region, we will either rebalance the entire colocation group (if it's the first region in the group to be executed on) or we'll exit early because `isRebalanceNecessary()` will return false (if any of the other regions in the group have already been executed on), so no additional work is being done at the region level.

The design for this class was originally much closer to what is seen in `RebalanceFactoryImpl`, but following feedback on the RFC in which several people requested that the `start()` method return a `CompletableFuture` rather than a separate Operation class, this approach was taken. I agree that it would be better to have both operations use the same logic, but `RebalanceOperation` and `RebalanceFactory` are both public APIs and so can't be easily changed, and there was a strong agreement on the RFC to use a `CompletableFuture` rather than the existing approach, so I'm not sure it's possible to make everyone happy here.